other_worker/location.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python
+
+import requests
+import logging
+import json
+
+from sqlalchemy import text
+from server import celery_app, db
+from server.models.location import Location
+from share.location import LocationExternalSourceEnum, osm_extension_fields
+
+QUERY_TIMEOUT = 2.0
+DEFAULT_COUNTRY_CODE = 'KE'
+VALID_OSM_ENTRY_TYPES =  ['village', 'suburb', 'address29', 'administrative', 'residential']
+
+logg = logging.getLogger(__file__)
+
+@celery_app.task()
+def search_name_from_osm(name, country=DEFAULT_COUNTRY_CODE):
+    url = 'https://nominatim.openstreetmap.org/search?format=json&dedupe=1&country={}&q={}'.format(country, name)
+
+    try:
+        response = requests.get(url, timeout=QUERY_TIMEOUT)
+    except requests.exceptions.Timeout:
+        logg.warning('request timeout to openstreetmap; {}:{}'.format(country, name))
+        # TODO: re-insert task
+        return
+
+    if response.status_code != 200:
+        logg.warning('failed request to openstreetmap; {}:{}'.format(country, name))
+        return
+
+    response_json = json.loads(response.text)
+    logg.debug(response_json)
+
+    # identify a suitable record among those returned
+    next_place_id = 0
+    for place in response_json:
+        if place['type'] in VALID_OSM_ENTRY_TYPES:
+            next_place_id = place['place_id']
+    if next_place_id == 0:
+        logg.debug('no suitable record found in openstreetmap for {}:{}'.format(country, name))
+        return
+
+    # get details of place, and recurse parents until top level is reached or found in db
+    url = 'https://nominatim.openstreetmap.org/details?format=json&linkedplaces=1&place_id=' 
+    locations = []
+    cache_index = -1
+    while next_place_id != 0:
+
+        sql = text('SELECT location_id FROM location_external WHERE external_reference ->> \'place_id\' = \'{}\''.format(next_place_id))
```

Comment:
text executes close to raw SQL - perhaps use bound and typed params to ensure we don't get injected, especially given next_place_id comes from an external api?
https://docs.sqlalchemy.org/en/13/core/tutorial.html#specifying-bound-parameter-behaviors